File: python/src/azoth/eos/reference/liquid_viscosity_pure.py
"""``eos.liquid_viscosity_pure`` - one component's pure-liquid viscosity, from the LIQVISC
correlation.

Spec: ``specs/calcs/eos/liquid_viscosity_pure.toml``, which records the four polynomial forms, the
pressure correction and **the two NeqSim classes whose ladders disagree**.
"""

from __future__ import annotations

import math
from typing import Any

from azoth.core.errors import InvalidInputError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import LiquidViscosityPureResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning

CALC_ID = "eos.liquid_viscosity_pure"

#: The value NeqSim answers above a component's critical temperature, in cP.
ABOVE_CRITICAL_CP = 0.5
#: The value it answers for a component with no LIQVISC model, in cP.
NO_MODEL_CP = 0.7
#: NeqSim's value is in cP; the port reports Pa*s.
CP_TO_PA_S = 1.0e-3

LADDERS = ("common_phase", "liquid")


def pressure_correction(
    temperature: float, pressure: float, tc: float, pc: float, omega: float
) -> float:
    """``getViscosityPressureCorrection``: NeqSim's own four-coefficient form."""
    reduced_t = temperature / tc
    if reduced_t > 1.0:
        return 1.0
    delta_pr = pressure / pc
    a = 0.9991 - (4.674e-4 / (1.0523 * math.pow(reduced_t, -0.03877) - 1.0513))
    d = (0.3257 / math.pow(1.0039 - reduced_t**2.573, 0.2906)) - 0.2086
    c = (
        -0.07921
        + 2.1616 * reduced_t
        - 13.4040 * reduced_t**2
        + 44.1706 * reduced_t**3
        - 84.8291 * reduced_t**4
        + 96.1209 * reduced_t**5
        - 59.8127 * reduced_t**6
        + 15.6719 * reduced_t**7
    )
    return (1.0 + d * math.pow(delta_pr / 2.118, a)) / (1.0 + c * omega * delta_pr)


def liquid_viscosity_pure(
    form: str,
    model: int,
    l1: float,
    l2: float,
    l3: float,
    l4: float,
    Tc: Q,
    Pc: Q,
    omega: float,
    T: Q,
    P: Q,
) -> LiquidViscosityPureResult:
    """One component's pure-liquid viscosity, from its LIQVISC correlation.

    ``form`` picks which of NeqSim's two ladders runs: ``common_phase``, whose LIQVISC model 2
    branch is empty so the component comes out at zero, or ``liquid``, which implements it.
    A phase gets whichever class its own viscosity model extends, and both are on the rate-based
    column's path.

    Args:
        form: ``common_phase`` or ``liquid``.
        model: the LIQVISC model index, 1 to 4; anything else answers the 0.7 cP sentinel.
        l1: ``LIQVISC1``.
        l2: ``LIQVISC2``.
        l3: ``LIQVISC3``, read by model 3 only.
        l4: ``LIQVISC4``, read by model 3 only.
        Tc: the component's critical temperature.
        Pc: its critical pressure.
        omega: its acentric factor.
        T: absolute temperature.
        P: absolute pressure.

    Returns:
        The component's pure-liquid viscosity, in ``Pa*s``.

    Raises:
        InvalidInputError: for a ``form`` that names neither ladder.
        OutOfRangeError: if ``T``, ``Tc`` or ``Pc`` is not positive.

    Example:
        >>> import azoth
        >>> q = azoth.ureg.Quantity
        >>> r = liquid_viscosity_pure(
        ...     "liquid", 3, -27.952757828, 4665.22592993, 0.052323342, -3.8356e-5,
        ...     q(647.3, "K"), q(22089000.0, "Pa"), 0.344,
        ...     q(313.15, "K"), q(50.0e5, "Pa"))
        >>> round(r.mu.magnitude, 18)
        0.0006528922494381046
    """
    if form not in LADDERS:
        raise InvalidInputError(
            "form",
            f"`{form}` is not one of the two ladders: {', '.join(LADDERS)}",
        )

    from azoth._registry_gen import spec as _spec_for

    spec: dict[str, Any] = _spec_for(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    values = {
        "Tc": input_to_si(spec, "Tc", Tc),
        "Pc": input_to_si(spec, "Pc", Pc),
        "T": input_to_si(spec, "T", T),
        "P": input_to_si(spec, "P", P),
    }
    apply_checks(checks.on_input, values.get, warnings)

    temperature = values["T"]
    if temperature > values["Tc"]:
        uncorrected = ABOVE_CRITICAL_CP
    elif model == 1:
        uncorrected = l1 * temperature**l2
    elif model == 2:
        # **The one branch the two ladders disagree on.**
        uncorrected = math.exp(l1 + l2 / temperature) if form == "liquid" else 0.0
    elif model == 3:
        uncorrected = math.exp(l1 + l2 / temperature + l3 * temperature + l4 * temperature**2)
    elif model == 4:
        uncorrected = 10 ** (l1 * (1.0 / temperature - 1.0 / l2))
    else:
        uncorrected = NO_MODEL_CP

    mu_cp = (
        uncorrected
        * (pressure_correction(temperature, values["P"], values["Tc"], values["Pc"], omega) + 1.0)
        / 2.0
    )
    mu = mu_cp * CP_TO_PA_S

    apply_checks(checks.derived, lambda name: mu if name == "mu" else None, warnings)

    return LiquidViscosityPureResult(mu=from_si(mu, "Pa*s"), warnings=tuple(warnings))

File: crates/azoth-eos/src/liquid_viscosity_pure.rs
//! `eos.liquid_viscosity_pure` - one component's pure-liquid viscosity, from the LIQVISC
//! correlation.
//!
//! Spec: `specs/calcs/eos/liquid_viscosity_pure.toml`, which records the four polynomial forms,
//! the pressure correction and **the two NeqSim classes whose ladders disagree**.

use azoth_core::units::{Pressure, ThermodynamicTemperature, pascal_seconds};
use azoth_core::{Result, apply_checks};

use crate::results::LiquidViscosityPureResult;
use crate::spec_gen;

/// The value NeqSim answers above a component's critical temperature, in cP.
const ABOVE_CRITICAL_CP: f64 = 0.5;
/// The value it answers for a component with no LIQVISC model, in cP.
const NO_MODEL_CP: f64 = 0.7;
/// A very small viscosity would divide; NeqSim clamps the *diffusivity's* use of this to 0.01 cP,
/// and the value itself is left as it comes out.
const CP_TO_PA_S: f64 = 1.0e-3;

/// **Which of NeqSim's two ladders to use.** They differ in exactly one branch.
///
/// `commonphasephysicalproperties.viscosity.Viscosity` leaves LIQVISC model 2's branch **empty**,
/// so a component with that model comes out at `0.0` cP and only the pressure correction scales
/// it; `liquidphysicalproperties.viscosity.Viscosity` implements it as `exp(l1 + l2/T)`. A phase
/// gets whichever class its own viscosity model extends, and both are on the rate-based column's
/// path: an oil phase's `PFCTViscosityMethodHeavyOil` inherits the first, an aqueous phase's
/// "polynom" model *is* the second.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LiquidViscosityLadder {
    /// `Viscosity` in `commonphasephysicalproperties` - model 2 answers nothing.
    CommonPhase,
    /// `Viscosity` in `liquidphysicalproperties` - every model answers.
    Liquid,
}

impl LiquidViscosityLadder {
    /// The spec's spelling.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::CommonPhase => "common_phase",
            Self::Liquid => "liquid",
        }
    }
}

impl std::str::FromStr for LiquidViscosityLadder {
    type Err = String;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        match s {
            "common_phase" => Ok(Self::CommonPhase),
            "liquid" => Ok(Self::Liquid),
            other => Err(format!(
                "`{other}` is not one of the two ladders: common_phase or liquid"
            )),
        }
    }
}

/// One component's pure-liquid viscosity, from its LIQVISC correlation.
///
/// The four models are NeqSim's own: `1` a power law, `2` an Arrhenius pair, `3` a four-term
/// exponential and `4` the Yaws form. `l1..l4` are the component's `LIQVISC1..4` columns, `tc` and
/// `pc` its critical point and `omega` its acentric factor, which the pressure correction reads.
///
/// The result is in Pa*s, where NeqSim's own value is in cP.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `T`, `Tc` or `Pc` is not positive.
///
/// # Example
/// ```
/// use azoth_core::units::{kelvins, pascals};
/// use azoth_eos::liquid_viscosity_pure::{LiquidViscosityLadder, liquid_viscosity_pure};
///
/// // Water at 313.15 K and 50 bar, the CO2 absorber's aqueous phase.
/// let r = liquid_viscosity_pure(
///     LiquidViscosityLadder::Liquid, 3,
///     -27.952757828, 4665.22592993, 0.052323342, -3.8356e-5,
///     kelvins(647.3), pascals(22089000.0), 0.344,
///     kelvins(313.15), pascals(50.0e5),
/// )?;
/// assert!((r.mu.value - 6.528922494381046e-4).abs() < 1e-18);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
#[allow(non_snake_case)] // `T`, `P`, `Tc`, `Pc` are the symbols in the correlation
#[allow(clippy::too_many_arguments)] // The signature is the spec's declared inputs.
pub fn liquid_viscosity_pure(
    form: LiquidViscosityLadder,
    model: u32,
    l1: f64,
    l2: f64,
    l3: f64,
    l4: f64,
    Tc: ThermodynamicTemperature,
    Pc: Pressure,
    omega: f64,
    T: ThermodynamicTemperature,
    P: Pressure,
) -> Result<LiquidViscosityPureResult> {
    let spec = &spec_gen::LIQUID_VISCOSITY_PURE_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "Tc" => Some(Tc.value),
            "Pc" => Some(Pc.value),
            "T" => Some(T.value),
            "P" => Some(P.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let uncorrected_cp = if T.value > Tc.value {
        ABOVE_CRITICAL_CP
    } else {
        match (form, model) {
            (_, 1) => l1 * T.value.powf(l2),
            // **Model 2 is the one branch the two ladders disagree on.**
            (LiquidViscosityLadder::Liquid, 2) => (l1 + l2 / T.value).exp(),
            (LiquidViscosityLadder::CommonPhase, 2) => 0.0,
            (_, 3) => (l1 + l2 / T.value + l3 * T.value + l4 * T.value.powi(2)).exp(),
            (_, 4) => 10f64.powf(l1 * (1.0 / T.value - 1.0 / l2)),
            _ => NO_MODEL_CP,
        }
    };

    let mu_cp = uncorrected_cp
        * (pressure_correction(T.value, P.value, Tc.value, Pc.value, omega) + 1.0)
        / 2.0;
    let mu = mu_cp * CP_TO_PA_S;

    apply_checks(
        spec.derived_checks(),
        |name| (name == "mu").then_some(mu),
        &mut warnings,
    )?;

    Ok(LiquidViscosityPureResult {
        mu: pascal_seconds(mu),
        warnings,
    })
}

/// `getViscosityPressureCorrection`: NeqSim's own four-coefficient form.
///
/// The Lucas pressure term, with the reduced temperature in the numerator of each exponential's
/// coefficient and the acentric factor in the denominator.
#[must_use]
pub fn pressure_correction(temperature: f64, pressure: f64, tc: f64, pc: f64, omega: f64) -> f64 {
    let reduced_t = temperature / tc;
    if reduced_t > 1.0 {
        return 1.0;
    }
    let delta_pr = pressure / pc;
    let a = 0.9991 - (4.674e-4 / (1.0523 * reduced_t.powf(-0.03877) - 1.0513));
    let d = (0.3257 / (1.0039 - reduced_t.powf(2.573)).powf(0.2906)) - 0.2086;
    let c = -0.07921 + 2.1616 * reduced_t - 13.4040 * reduced_t.powi(2)
        + 44.1706 * reduced_t.powi(3)
        - 84.8291 * reduced_t.powi(4)
        + 96.1209 * reduced_t.powi(5)
        - 59.8127 * reduced_t.powi(6)
        + 15.6719 * reduced_t.powi(7);
    (1.0 + d * (delta_pr / 2.118).powf(a)) / (1.0 + c * omega * delta_pr)
}
